### ISSUE TITLE
Allow Hardentool to disable Microsoft Recall.

### ISSUE THREAD

**Issue by johnnymatthews** [hardentools/hardentools#131]
Sorry if this doesn't fit within the pervue of this repo, it's just something I figured might be of interest to folks looking to reduce their Windows attack vectors.

From the [Microsoft Docs](https://learn.microsoft.com/en-us/windows/client-management/manage-recall#configure-policies-for-recall):

> Recall allows you to search across time to find the content you need. Just describe how you remember it, and Recall retrieves the moment you saw it. Recall takes snapshots of your screen and stores them in a timeline. Snapshots are taken every five seconds while content on the screen is different from the previous snapshot. Snapshots are locally stored and locally analyzed on your PC. Recall's analysis allows you to search for content, including both images and text, using natural language.

The security of this feature is still a big question, but disabling it outright is a surefire way to mitigate any security holes Recall has.

**Comment by johnnymatthews** [comment]
According to the Microsoft Docs, the policy to disable is **User Configuration** > **Administrative Templates** > **Windows Components** > **Windows AI** > **Turn off saving snapshots for Windows**. I'm not super familiar with how this repo is organized, or Golang for that matter, but I'll do a little digging and create a PR if I can.

**Comment by obsti8383** [comment]
Hi @johnnymatthews , I think this exactly fits for our intended target audience. So would be great to have that feature disabled by Hardentools as soon as Microsoft goes into rollout. Do you have access to the Microsoft Insider community versions to be able to test (reference: https://blogs.windows.com/windowsexperience/2024/06/07/update-on-the-recall-preview-feature-for-copilot-pcs/)?
Please don't hesitate to ask if you have any question regarding to organisation of the Golang stuff here. In general it is quite easy to implement if it is only a registry setting (sounds like it from your description above). lsa_protection.go is a good example, if it is only a single registry key. The struct created there then needs to be added to hardenSubjectsForUnprivilegedUsers or hardenSubjectsForPrivilegedUsers (depending if admin rights are needed) in global_vars.go. Thats it.

**Comment by johnnymatthews** [comment]
Hey @obsti8383! I took a very cursory look at this repo to see where I should begin, but got sidetracked. Since I'm not super familiar with Go, it'd take me a while to put a PR together. I think I'd enjoy doing that, _but_ I also don't want to hold up this feature getting added into Hardentools. So if yourself, or anyone else, would like to see this implemented quickly, feel free! I won't be upset if someone manages to close this issue before me 😅 .

> Do you have access to the Microsoft Insider community versions

Yeah, I do.

**Comment by johnnymatthews** [comment]
I'm not the project owner, so I can't speak to that. I never got around to implementing this feature -- Go is not my strong-point. The last release from this repo is from 2023, so it does seems that the owner has moved onto other things (which is completely understandable).

**Comment by obsti8383** [comment]
Hi @brrock, Hi @johnnymatthews project is still alive, no worries. But no big things to improve / adapt. I think since Microsoft Recall is now rolling out at least on hardware with the necessary "Neural Processing Units", it might make sense to implement that. However, since it is still disabled by default, question is if makes sense. I have to feeling that Microsoft might press users into activating it, so perhaps the answer should be "yes". Anyone of you have experience with the feature already? I don't have the necessary hardware yet.
I've taken a brief look at [https://learn.microsoft.com/en-us/windows/client-management/manage-recall](https://learn.microsoft.com/en-us/windows/client-management/manage-recall) and [https://www.dennis-sareika.de/artikel/technik/microsoft-recall-in-windows-11-version-24h2-deaktivieren](https://www.dennis-sareika.de/artikel/technik/microsoft-recall-in-windows-11-version-24h2-deaktivieren). My feeling is that the safest  way to disable Recall would be to actually uninstall the Windows feature with `Disable-WindowsOptionalFeature -Online -FeatureName "Recall" -Remove` or `Dism /Online /Disable-Feature /Featurename:Recall`. Before trying that we could check if it is active with `Dism /Online /Get-Featureinfo /Featurename:Recall`. But thats just theory and would need some testing with appropriate hardware. Does anyone of you have a system thats capable of running Recall?

**Comment by brrock** [comment]
👍👍

**Comment by obsti8383** [comment]
@brrock , @johnnymatthews : Please take a look at a first draft version here: https://github.com/hardentools/hardentools/releases/tag/v2.7-beta

Please be aware that you have to check the "Recall" feature in the expert settings, since I've made it optional for now (and perhaps will keep it that way, because it really takes long to activate/deactivate, at least as long as it is opt-in from Microsoft side)

If something does not work as expected, please provide a bug report including the hardentools.log file (it should do some verbose logging for the Recall feature.

**Comment by johnnymatthews** [comment]
Oh gosh I'm sorry! I've been AFK for a while. Let me get this tested.

**Comment by obsti8383** [comment]
> Oh gosh I'm sorry! I've been AFK for a while. Let me get this tested.

Great, thanks

**Comment by johnnymatthews** [comment]
Here's what I did:

1. Downloaded the package and ran it with Admin permissions.
1. Unhardened my settings (I didn't fancy reformatting my machine just yet).
1. Restarted Windows.
1. Booted up Hardentools GUI (again, with Admin permissions).
1. Selected **Show Expert Settings**.
1. Noticed that **Recall Windows Feature** was greyed-out. I'm assuming this is because I disabled it at some point using Chris Titus Tech's WinUtil tool.
1. Booted up said WinUtil tool:

	```shell
	iwr -useb https://christitus.com/win | iex
	```

1. Switched to the **Tweaks** tab, selected **Disable Recall** and clicked **Undo Selected Tweaks**.

That's as far as I got. The WinUtil tool seems to have difficutly re-adding Recall - which is probably a good thing but for our situation is a tiny roadblock.

I'm not sure a VM will work in this case. I'm gonna have to reformat this machine and start a fresh.

**Comment by johnnymatthews** [comment]
Minor problem: I booted up the GUI (with Admin privs) and entered **Expert Mode*, but the checkbox to set Recall was greyed out. This is a fresh installed of Windows 11.  Not sure what's happening.

<img width="758" height="503" alt="Image" src="https://github.com/user-attachments/assets/aa51db7b-13c2-4665-a86b-8d723f04dd4b" />

I'm gonna spin up a fresh VM to see what the deal is.

**Comment by brrock** [comment]
Do you need copilot+?

**Comment by obsti8383** [comment]
> Minor problem: I booted up the GUI (with Admin privs) and entered *_Expert Mode_, but the checkbox to set Recall was greyed out. This is a fresh installed of Windows 11. Not sure what's happening.
> <img alt="Image" width="758" height="503" src="https://private-user-images.githubusercontent.com/9611008/484758932-aa51db7b-13c2-4665-a86b-8d723f04dd4b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY5MjAxMzQsIm5iZiI6MTc1NjkxOTgzNCwicGF0aCI6Ii85NjExMDA4LzQ4NDc1ODkzMi1hYTUxZGI3Yi0xM2MyLTQ2NjUtYTg2Yi04ZDcyM2YwNGRkNGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MDNUMTcxNzE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDJkZDZjYTU0ZTJjZTkwMTYzY2ZhMGI0MDI2ZWM0YmRhYjEwZTY2MjBlNWRkM2RiOTI1ZWZkYTcwNTBmNTIzMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.aPaW7VEiH4eV1CEjt2RtMqobeN7xRKod0p2v53iU5uY">
> 
> I'm gonna spin up a fresh VM to see what the deal is.

Could you please check the contents of the hardentools.log file and also the output of the following command?

`Get-WindowsOptionalFeature -Online -FeatureName "Recall"`

If the status is `DisabledWithPayloadRemoved` then its not installed right now. 

Does your system have Bitlocker activated? 
Please also check if Recall is installed in the Windows Features.

**Comment by johnnymatthews** [comment]
```powershell
PS C:\Users\Johnny> Get-WindowsOptionalFeature -Online -FeatureName "Recall"


FeatureName      : Recall
DisplayName      : Recall
Description      : Recall application.
RestartRequired  : Possible
State            : DisabledWithPayloadRemoved
CustomProperties :
```

Hardentools.log file:

```plaintext
INFO: recall_feature.go:117: Recall: Executing Powershell.exe with command "Get-WindowsOptionalFeature -Online -FeatureName "Recall" | Select-Object -ExpandProperty State"
INFO: recall_feature.go:125: Recall: Powershell output for test of status was:
DisabledWithPayloadRemoved
INFO: recall_feature.go:129: Recall: Is hardened
```

**Comment by obsti8383** [comment]
@johnnymatthews Yes, State: DisabledWithPayloadRemoved says us that Recall is disabled. You would have to install/activate it first and then start hardentools again. "Recall: Is hardened" in the hardentools log also says us that hardentools has identified it as already hardened (not installed) and therefore the button is greyed out.
